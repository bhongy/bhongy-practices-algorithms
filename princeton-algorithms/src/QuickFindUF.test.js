// @flow
const { connected, union } = require('./QuickFindUF');

describe('QuickFindUF', () => {
  function createStore(n: number): number[] {
    return Array.from({ length: n }, (_, i) => i);
  }

  function applyUnions<T: number[]>(store: T): T {
    return [
      [4, 3],
      [3, 8],
      [6, 5],
      [9, 4],
      [2, 1],
      [5, 0],
      [7, 2],
    ].reduce((prev, [p, q]) => union(prev, p, q), store);
  }

  it('unions correctly', () => {
    const initial = createStore(10);
    expect(initial).toMatchSnapshot('Initial State');

    const unioned = applyUnions(initial);
    expect(unioned).toMatchSnapshot('After unions');
  });

  it('checks connetions correctly', () => {
    const s = applyUnions(createStore(10));

    expect(connected(s, 6, 0)).toBe(true);
    expect(connected(s, 8, 9)).toBe(true);

    expect(connected(s, 2, 0)).toBe(false);
    expect(connected(s, 3, 7)).toBe(false);
  });

  it('handles out of range indices', () => {
    [
      { fn: union, p: -1, q: 5 },
      { fn: union, p: 1, q: 12 },
      { fn: connected, p: -3, q: 12 },
      { fn: connected, p: 12, q: -3 },
    ].forEach(({ fn, p, q }) => {
      const s = applyUnions(createStore(10));
      expect(() => fn(s, p, q)).toThrowErrorMatchingSnapshot();
    });
  });
});
