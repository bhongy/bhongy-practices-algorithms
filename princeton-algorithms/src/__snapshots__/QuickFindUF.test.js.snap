// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuickFindUF handles out of range indices 1`] = `"id is out of bound. Expect between [0, 9] but received -1"`;

exports[`QuickFindUF handles out of range indices 2`] = `"id is out of bound. Expect between [0, 9] but received 12"`;

exports[`QuickFindUF handles out of range indices 3`] = `"id is out of bound. Expect between [0, 9] but received -3"`;

exports[`QuickFindUF handles out of range indices 4`] = `"id is out of bound. Expect between [0, 9] but received 12"`;

exports[`QuickFindUF unions correctly: After unions 1`] = `
Array [
  0,
  1,
  1,
  8,
  8,
  0,
  0,
  1,
  8,
  8,
]
`;

exports[`QuickFindUF unions correctly: Initial State 1`] = `
Array [
  0,
  1,
  2,
  3,
  4,
  5,
  6,
  7,
  8,
  9,
]
`;
